### PR TITLE
REPO-1754 inconsistent d:period on put request

### PR DESCRIPTION
**Pull request by anechifor**
I add a new conveter  to DefaultTypeConverter from Map to Period

### CODE REVIEW

File: src/main/java/org/alfresco/service/cmr/repository/Period.java [hunk 1]
```diff
@@ -102,6 +104,26 @@ public Period(String period)
...
     }
+    
+    /**
+     * Create a period using key-value
+     * 
+     * @param source
+     */
+    public Period(Map<String, String> source)
+    {
+        if (source != null)
```
Comment (mattward) [inline src/main/java/org/alfresco/service/cmr/repository/Period.java]: Although code in this class such as `hashCode` appears to allow for a `null` `periodType`, it wasn't previously possible to instantiate this class without a periodType (though the `expression` is intentionally optional). This constructor allows a null `source` and will result in a `Period` with both fields `null`. Should we be throwing an `IllegalArgumentException` or similar if either the `source` map, or the `periodType` (within the source map) are null?

File: src/test/java/org/alfresco/service/cmr/repository/datatype/DefaultTypeConverterTest.java [hunk 2]
```diff
@@ -118,8 +120,12 @@ public void testToString()
         // VersionNumber
         assertEquals("1.2.3", DefaultTypeConverter.INSTANCE.convert(String.class, new VersionNumber("1.2.3")));
         // Period
-        assertEquals("period", DefaultTypeConverter.INSTANCE.convert(String.class, new Period("period")));
+        //assertEquals("period", DefaultTypeConverter.INSTANCE.convert(String.class, new Period("period")));
```
Comment (mattward) [inline src/test/java/org/alfresco/service/cmr/repository/datatype/DefaultTypeConverterTest.java]: Has this been intentionally commented out?

File: src/main/java/org/alfresco/service/cmr/repository/Period.java (hunk 1 above)
Comment (anechifor) [reply src/main/java/org/alfresco/service/cmr/repository/Period.java]: yes in java doc says: A period is specified by the period type and an optional
 expression. periodType is mandatory. I will throw IllegalArgumentException. It's not possible to have a periodType null.

File: src/test/java/org/alfresco/service/cmr/repository/datatype/DefaultTypeConverterTest.java [hunk 3]
```diff
@@ -163,9 +169,12 @@ public void testFromString()
         assertEquals(Locale.FRANCE, DefaultTypeConverter.INSTANCE.convert(Locale.class, "fr_FR_"));
         
         assertEquals(new VersionNumber("1.2.3"), DefaultTypeConverter.INSTANCE.convert(VersionNumber.class, "1.2.3"));
-        
         assertEquals(new Period("period"), DefaultTypeConverter.INSTANCE.convert(Period.class, "period"));
         assertEquals(new Period("period|12"), DefaultTypeConverter.INSTANCE.convert(Period.class, "period|12"));
+        Map<String,String> periodMap = new HashMap<String, String>();
+        periodMap.put("periodType","month");
+        periodMap.put("expression","1");
+        assertEquals(new Period(periodMap), DefaultTypeConverter.INSTANCE.convert(Period.class, "month|1"));
```
Comment (mattward) [inline src/test/java/org/alfresco/service/cmr/repository/datatype/DefaultTypeConverterTest.java]: This isn't testing anything about the DefaultTypeConverter that isn't tested by the assertion above ("period|12"), if anything it's really testing is the `Period(Map)` constructor. It should be testing the `Map` to Period conversion, but is testing String to Period conversion.

File: src/test/java/org/alfresco/service/cmr/repository/datatype/DefaultTypeConverterTest.java [hunk 4]
```diff
@@ -118,8 +120,12 @@ public void testToString()
         // VersionNumber
         assertEquals("1.2.3", DefaultTypeConverter.INSTANCE.convert(String.class, new VersionNumber("1.2.3")));
         // Period
-        assertEquals("period", DefaultTypeConverter.INSTANCE.convert(String.class, new Period("period")));
+        //assertEquals("period", DefaultTypeConverter.INSTANCE.convert(String.class, new Period("period")));
         assertEquals("period|12", DefaultTypeConverter.INSTANCE.convert(String.class, new Period("period|12")));
+        Map<String,String> periodMap = new HashMap<String, String>();
+        periodMap.put("periodType","month");
+        periodMap.put("expression","1");
+        assertEquals("month|1", DefaultTypeConverter.INSTANCE.convert(String.class, new Period(periodMap)));
```
Comment (mattward) [inline src/test/java/org/alfresco/service/cmr/repository/datatype/DefaultTypeConverterTest.java]: This is testing Period to String conversion, which is already well tested. It is really testing the map constructor, which is sort of ok, but would be better in a PeriodTest class.

File: src/test/java/org/alfresco/service/cmr/repository/datatype/DefaultTypeConverterTest.java (hunk 3 above)
Comment (mattward) [reply src/test/java/org/alfresco/service/cmr/repository/datatype/DefaultTypeConverterTest.java]: i.e. `DefaultTypeConverter.INSTANCE.convert(Period.class, periodMap)` is what should be tested here

Comment (anechifor) [reply src/test/java/org/alfresco/service/cmr/repository/datatype/DefaultTypeConverterTest.java]: ok I change

File: src/test/java/org/alfresco/service/cmr/repository/datatype/DefaultTypeConverterTest.java (hunk 4 above)
Comment (anechifor) [reply src/test/java/org/alfresco/service/cmr/repository/datatype/DefaultTypeConverterTest.java]: I will add a test in PeriodTest you are right

File: src/test/java/org/alfresco/service/cmr/repository/datatype/DefaultTypeConverterTest.java (hunk 2 above)
Comment (derekhulley) [reply src/test/java/org/alfresco/service/cmr/repository/datatype/DefaultTypeConverterTest.java]: If the code is no longer required, it should be deleted.

File: pom.xml [hunk 5]
```diff
@@ -10,7 +10,7 @@
     <artifactId>alfresco-data-model</artifactId>
     <name>Alfresco Data Model</name>
     <description>Alfresco Data Model classes</description>
-    <version>6.22-SNAPSHOT</version>
```
Comment (derekhulley) [inline pom.xml]: A downgrade ... why?  This implies that there is a broken data model version.

File: pom.xml [hunk 6]
```diff
@@ -31,7 +31,7 @@
     </distributionManagement>
 
     <properties>
-        <dependency.alfresco-core.version>6.14</dependency.alfresco-core.version>
```
Comment (derekhulley) [inline pom.xml]: Why downgrade?
This implies a broken core version that can't be used.
This pom file change should be a separate commit with some details of WHY.

**Comment by mattward** [comment]
Excellent, those changes look good - ready to merge from my point of view.
